### DISCUSSION TITLE
Invoking method is an array argument

### DISCUSSION THREAD

**Discussion by shiirish** [Kistler-Group/sdbus-cpp#206] [Q&A] [answered]
Hi all,

How can I invoke the `CreateNetwork` method on the 'org.bluez.mesh.Network1' interface? I've skimmed the D-Bus XML using D-Feet and generated the proxy help file. When I invoke the method, I get the following:

> org.freedesktop.DBus.Error.InvalidArgs with message Failed to create method call (Invalid argument)

The XML method definition is:
```xml
[...]
<method name="CreateNetwork">
  <arg name="app" type="o" direction="in"/>
  <arg name="uuid" type="ay" direction="in"/>
</method>
[...]
```

And the generated code is:
```c++
void CreateNetwork(const sdbus::ObjectPath& app, const std::vector<uint8_t>& uuid)
{
    proxy_.callMethod("CreateNetwork").onInterface(INTERFACE_NAME).withArguments(app, uuid);
}
```
I'm invoking it as:
```c++
static std::string serviceName = "com.example.BLEMesh";
static std::string objectPath = "/com/example/blemesh";

// Create connection
auto connection = sdbus::createSystemBusConnection(serviceName);
// Create proxy w/ connection
auto networkProxy = new org::bluez::mesh::Network(*connection, "org.bluez.mesh", "org.bluez.mesh.Network1");

// Enter the event loop
connection->enterEventLoopAsync();

// Sleep to let things spin up (?)
sleep(1);

// Generate GUID, get ObjectPath obj
std::vector<uint8_t> uuid = generateUUID();
sdbus::ObjectPath objectPathObj(objectPath);

// Invoke method
networkProxy->CreateNetwork(objectPathObj, uuid);
```
I wrote a C program using the `systemd` library to make sure I was addressing the method correctly. I was able to make that work, and noticed that the C code calls `sd_bus_message_append_array` to set up the message. I also noticed that `sd_bus_message_append_array` is not referenced in sdbus-c++. I'm guessing that there is a different method of loading the data into the message, but I'm not able to find anything that has worked for me yet.

Any help is appreciated!

Thanks,

JD Scott

**Comment by dleeds-cpi** [comment, marked as answer]
Hi @shiirish, here are a few notes on usage:
1. You don't need to specify a service name for your D-Bus connection, because you are a client. You only need a well-known-name if someone is calling into you.
2. You haven't included all your code here, but I assume your `org::bluez::mesh::Network` proxy class takes three parameters: an `IConnection` reference, a D-Bus service name, and a D-Bus object path. You are passing it a D-Bus interface for the third parameter.

**Reply by shiirish** [reply]
@dleeds-cpi Thanks for the quick response. I was literally in the middle of admitting that I was not passing the correct value as the object path. As for 1, the Mesh service of BlueZ requires a service be made available and identified for `CreateNetwork`. That is the next step for me.

Again, I appreciate the help!
